cmake/modules/FindGINKGO.cmake

```diff
@@ -19,54 +19,53 @@
 # This module exports
 #
 #   GINKGO_INCLUDE_DIRS
+#   GINKGO_INTERFACE_LINK_FLAGS
 #
 
 SET(GINKGO_DIR "" CACHE PATH "An optional hint to a GINKGO installation")
 SET_IF_EMPTY(GINKGO_DIR "$ENV{GINKGO_DIR}")
 
-DEAL_II_FIND_LIBRARY(GINKGO_LIBRARY
-  NAMES ginkgo
-  HINTS ${GINKGO_DIR}
-  PATH_SUFFIXES
-    lib${LIB_SUFFIX} lib64 lib
-    # This is a hint, isn't it?
-    build/${CMAKE_CXX_PLATFORM_ID}-${CMAKE_SYSTEM_PROCESSOR}/libginkgo
-  )
-DEAL_II_FIND_LIBRARY(GINKGO_REFERENCE_LIBRARY
-  NAMES ginkgo_reference
-  HINTS ${GINKGO_DIR}
-  PATH_SUFFIXES
-    lib${LIB_SUFFIX} lib64 lib
-    # This is a hint, isn't it?
-    build/${CMAKE_CXX_PLATFORM_ID}-${CMAKE_SYSTEM_PROCESSOR}/libginkgo_reference
-  )
-DEAL_II_FIND_LIBRARY(GINKGO_OMP_LIBRARY
-  NAMES ginkgo_omp
-  HINTS ${GINKGO_DIR}
-  PATH_SUFFIXES
-    lib${LIB_SUFFIX} lib64 lib
-    # This is a hint, isn't it?
-    build/${CMAKE_CXX_PLATFORM_ID}-${CMAKE_SYSTEM_PROCESSOR}/libginkgo_omp
-  )
-DEAL_II_FIND_LIBRARY(GINKGO_CUDA_LIBRARY
-  NAMES ginkgo_cuda
-  HINTS ${GINKGO_DIR}
-  PATH_SUFFIXES
-    lib${LIB_SUFFIX} lib64 lib
-    # This is a hint, isn't it?
-    build/${CMAKE_CXX_PLATFORM_ID}-${CMAKE_SYSTEM_PROCESSOR}/libginkgo_cuda
+#
+# Save and restore the ${CMAKE_MODULE_PATH} variable. The Ginkgo project
+# configuration unfortunately overrides the variable which causes
+# subsequent configuration to fail.
+#
+SET(_cmake_module_path ${CMAKE_MODULE_PATH})
+FIND_PACKAGE(Ginkgo
+  HINTS ${GINKGO_DIR} ${Ginkgo_DIR} $ENV{Ginkgo_DIR}
   )
+SET(CMAKE_MODULE_PATH ${_cmake_module_path})
 
-DEAL_II_FIND_PATH(GINKGO_INCLUDE_DIR ginkgo/ginkgo.hpp
-  HINTS ${GINKGO_DIR}
-  PATH_SUFFIXES include
-  )
+#
+# Cosmetic clean up: Let's remove all variables beginning with "GINKGO_"
+# that are actually not used during configuration but show up in
+# detailed.log
+#
+unset(GINKGO_CXX_COMPILER)
+
+#
+# We'd like to have the full library names but the Ginkgo package only
+# exports a list with short names. So check again for every lib and store
+# the full path:
+#
+SET(_libraries "")
+FOREACH(_library ${GINKGO_INTERFACE_LINK_LIBRARIES})
+  LIST(APPEND _libraries GINKGO_LIBRARY_${_library})
+  DEAL_II_FIND_LIBRARY(GINKGO_LIBRARY_${_library}
+    NAMES ${_library}
+    HINTS ${GINKGO_INSTALL_LIBRARY_DIR}
+    NO_DEFAULT_PATH
+    NO_CMAKE_ENVIRONMENT_PATH
+    NO_CMAKE_PATH
+    NO_SYSTEM_ENVIRONMENT_PATH
+    NO_CMAKE_SYSTEM_PATH
+    NO_CMAKE_FIND_ROOT_PATH
+    )
+ENDFOREACH()
```

Comment:
@pratikvn Ideally, I would like to recover the full library paths from the exported import targets. That's unfortunately a bit tricky. On my Gentoo Linux machine for example I have in `GinkgoTargets.cmake`:
```
# Import target "Ginkgo::ginkgo_cuda" for configuration "Gentoo"
set_property(TARGET Ginkgo::ginkgo_cuda APPEND PROPERTY IMPORTED_CONFIGURATIONS GENTOO)
set_target_properties(Ginkgo::ginkgo_cuda PROPERTIES
  IMPORTED_LOCATION_GENTOO "/var/tmp/portage/sci-libs/ginkgo-1.1.1-r1/work/ginkgo-1.1.1_build/core/device_hooks/libginkgo_cuda.so.1.1.1"
  IMPORTED_SONAME_GENTOO "libginkgo_cuda.so.1.1.1"
  )
```
Now querying for the compile configuration and querying the `IMPORTED_LOCATION_GENTOO` (or `RELEASE`, or `DEBUG`) is ugly but doable. The real issue is that this doesn't seem to be the installed target location but an exported target during compilation, i.e., a bug in the Ginkgo CMake configuration.

So for now, invoking `FIND_LIBRARY` again (and forcing the search path to `${GINKGO_INSTALL_LIBRARY_DIR}`) seems to be the best option - we do the same for Trilinos.

Reply:
I see. It looks like we fixed this in a later version of Ginkgo (https://github.com/ginkgo-project/ginkgo/pull/414). But for backward compatibility, your solution looks good to me!

Reply:
By the way, we can also enforce that deal.ii only supports Ginkgo v1.2.0 versions and above. We are okay with that and would probably prefer it as well.